### PR TITLE
PR: Implement support for "HCL" colourspace.

### PR DESCRIPTION
**Pull request by Saransh-cpp**
Issue #564 

This is a new PR as the last one had some issues (discussed in #776).

## Additions
1. Functions for converting RGB colourspace to HCL colourspace and vice-versa.
2. Imports in the required files.
3. Tests for `RGB_to_HCL` and `HCL_to_RGB` conversion.
4. RGB_to_HCL and HCL_to_RGB in colour/graph/conversion.py.
5. Examples for the added conversions. 
6. Documentation entries added to docs/colour.models.rst

## Things left to add 
(as instructed by @KelSolaar )
Took me some time but added everything.

### CODE REVIEW

**Comment by KelSolaar** [comment]
Hi @Saransh-cpp,

Let us know when it is ready to review!

Cheers,

Thomas

**Comment by Saransh-cpp** [comment]
@KelSolaar yes, it is ready to be reviewed. 

Thanks!

File: colour/__init__.py [hunk 1]
```diff
@@ -131,7 +131,7 @@
     matrix_YCbCr, normalised_primary_matrix, oetf, oetf_inverse, offset_YCbCr,
     ootf, ootf_inverse, primaries_whitepoint, sRGB_to_XYZ, uv_to_Luv,
     uv_to_UCS, xyY_to_XYZ, xyY_to_xy, xy_to_Luv_uv, xy_to_UCS_uv, xy_to_XYZ,
-    xy_to_xyY)
+    xy_to_xyY, RGB_to_HCL, HCL_to_RGB)
```
Comment (KelSolaar) [inline colour/__init__.py:134]: No biggie, but they should ideally be added in alphabetical order, with uppercasing taking precedence.

File: colour/__init__.py [hunk 2]
```diff
@@ -292,7 +292,8 @@ def __getattr__(self, attribute):
     'matrix_YCbCr', 'normalised_primary_matrix', 'oetf', 'oetf_inverse',
     'offset_YCbCr', 'ootf', 'ootf_inverse', 'primaries_whitepoint',
     'sRGB_to_XYZ', 'uv_to_Luv', 'uv_to_UCS', 'xyY_to_XYZ', 'xyY_to_xy',
-    'xy_to_Luv_uv', 'xy_to_UCS_uv', 'xy_to_XYZ', 'xy_to_xyY'
+    'xy_to_Luv_uv', 'xy_to_UCS_uv', 'xy_to_XYZ', 'xy_to_xyY', 'RGB_to_HCL',
+    'HCL_to_RGB'
```
Comment (KelSolaar) [inline colour/__init__.py:296]: Likewise for the ordering.

File: colour/examples/models/examples_cylindrical.py [hunk 3]
```diff
@@ -40,3 +40,18 @@
 print(colour.HSL_to_RGB(HSL))
 
 print('\n')
+
+message_box(('Converting to "HCL" colourspace from given "RGB" colourspace '
```
Comment (KelSolaar) [inline colour/examples/models/examples_cylindrical.py:44]: Nice!

File: colour/graph/conversion.py [hunk 4]
```diff
@@ -44,7 +44,8 @@
     XYZ_to_hdr_CIELab, XYZ_to_hdr_IPT, XYZ_to_sRGB, XYZ_to_xy, XYZ_to_xyY,
     YCbCr_to_RGB, YCoCg_to_RGB, YcCbcCrc_to_RGB, cctf_decoding, cctf_encoding,
     hdr_CIELab_to_XYZ, hdr_IPT_to_XYZ, sRGB_to_XYZ, uv_to_Luv, uv_to_UCS,
-    xyY_to_XYZ, xyY_to_xy, xy_to_Luv_uv, xy_to_UCS_uv, xy_to_XYZ, xy_to_xyY)
+    xyY_to_XYZ, xyY_to_xy, xy_to_Luv_uv, xy_to_UCS_uv, xy_to_XYZ, xy_to_xyY,
+    RGB_to_HCL, HCL_to_RGB)
```
Comment (KelSolaar) [inline colour/graph/conversion.py:48]: Same than for the imports for the ordering.

File: colour/models/rgb/cylindrical.py [hunk 5]
```diff
@@ -31,6 +33,11 @@
     Interactive Techniques - SIGGRAPH "78, 12-19. doi:10.1145/800248.807361
 -   :cite:`Wikipedia2003` : Wikipedia. (2003). HSL and HSV. Retrieved
     September 10, 2014, from http://en.wikipedia.org/wiki/HSL_and_HSV
+-   :cite:`Sarifuddin2005ANP` : M. SarifuddinA (2005) New Perceptually Uniform
```
Comment (KelSolaar) [inline colour/models/rgb/cylindrical.py:36]: Here are the citation key and bibliography:

- `Sarifuddin2005`
- `Sarifuddin, M., & Missaoui, R. (2005). A New Perceptually Uniform Color Space with Associated Color Similarity Measure for ContentBased Image and Video Retrieval.`

File: colour/models/rgb/cylindrical.py [hunk 6]
```diff
@@ -31,6 +33,11 @@
...
+    Color Space with Associated Color Similarity Measure for Content-Based
+    Image and Video Retrieval
+-   :cite:`Wikipedia2015`, Wikipedia. (created 2015). HCL color space,
```
Comment (KelSolaar) [inline colour/models/rgb/cylindrical.py:39]: Bibliography for this one:

- `Wikipedia. (2015). HCL color space. Retrieved April 4, 2021, from https://en.wikipedia.org/wiki/HCL_color_space`

File: colour/models/rgb/cylindrical.py [hunk 7]
```diff
@@ -336,3 +343,184 @@ def H_to_RGB(vi, vj, vH):
...
+
+def RGB_to_HCL(RGB, gamma=3, Y_0=100):
+    """
+    Converts from *RGB* colourspace to *HCL* colourspace.
+
+    Parameters
+    ----------
+    RGB : array_like
+        *RGB* colourspace array.
+    gamma : double, optional
```
Comment (KelSolaar) [inline colour/models/rgb/cylindrical.py:356]: Should be `numeric, optional` and we need parameter description likewise for `Y_0`.

File: colour/models/rgb/cylindrical.py [hunk 8]
```diff
@@ -336,3 +343,184 @@ def H_to_RGB(vi, vj, vH):
...
+
+def HCL_to_RGB(HCL, gamma=3, Y_0=100):
+    """
+    Converts from *HCL* colourspace to *RGB* colourspace.
+
+    Parameters
+    ----------
+    HCL : array_like
+        *HCL* colourspace array.
+    gamma : double, optional
```
Comment (KelSolaar) [inline colour/models/rgb/cylindrical.py:435]: Ditto for the parameters docs! :)

**Comment by KelSolaar** [comment]
Thanks and great work @Saransh-cpp! Only a few minor notes and it is ready to merge.

**Comment by Saransh-cpp** [comment]
Thanks, @KelSolaar, please let me know the required changes and I will push them in the next commits.

**Review by KelSolaar — COMMENTED** [review]
Here are the review notes!

File: colour/models/rgb/cylindrical.py (hunk 7 above)
Comment (Saransh-cpp) [reply colour/models/rgb/cylindrical.py:356]: Oh yes, somehow missed the Y_0 parameter,

**Comment by Saransh-cpp** [comment]
@KelSolaar thank you so much for the review! I have made the necessary changes in my local files but I am again having some problems with committing the files. The error - 
```terminal
flake8...................................................................Passed
yapf.....................................................................Failed
- hook id: yapf
- files were modified by this hook
[INFO] Restored changes from C:\Users\Saransh/.cache\pre-commit\patch1617529106.
```

The file mentioned in the error - 
```txt
diff --git a/pyproject.toml b/pyproject.toml
index 8008ea47..09c8055b 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -91,6 +91,7 @@ toml = "*"
 tqdm = "*"
 twine = "*"
 yapf = "0.23"
+black = {version = "^20.8b1", allow-prereleases = true}
 
 [tool.poetry.extras]
 development = [
```

Could you please guide me as to what can be done to fix this?

**Comment by KelSolaar** [comment]
It seems like that for some reason you have installed `black` in the `colour`'s Poetry virtual environment. We use `Yapf` by the way.

**Comment by Saransh-cpp** [comment]
Yes, I frequently use black for formatting purposes, that might've installed it. I will try to remove it from there.

**Comment by Saransh-cpp** [comment]
@KelSolaar, I did uninstall black from the virtual environment located at `C:\Users\Saransh\AppData\Local\pypoetry\Cache\virtualenvs\colour-KBwigDaU-py3.9` but the yapf test still fails, am I required to remove it from somewhere else too?

**Comment by KelSolaar** [comment]
Would you mind pasting the full log as I don't see why `yapf` would update the `pyproject.toml` file. How did you setup the pre-commit hooks?

**Comment by Saransh-cpp** [comment]
Could you please confirm what logs am I supposed to paste here? The pre-commit hooks were installed using 
```terminal 
poetry run pre-commit install
```
I just checked and the .toml file was edited with an addition of "black", I have removed it from there too.

**Comment by KelSolaar** [comment]
So, just to be clear, it is failing when you are trying to commit?

**Comment by Saransh-cpp** [comment]
> So, just to be clear, it is failing when you are trying to commit?

I just tried to commit again to reproduce the failure and surprisingly it passed :). Sorry for all the inconveniences, I will push the commit. Thanks for helping me out!

**Comment by Saransh-cpp** [comment]
@KelSolaar I have made the mentioned changes, please let me know if any other changes are required. Thanks!

**Comment by KelSolaar** [comment]
Thanks @Saransh-cpp, it is in the box! :)

**Pull request merged**